Mathlib/NumberTheory/Transcendental/AnalyticPart.lean

```diff
@@ -0,0 +1,320 @@
+              left
+              rw [mul_assoc, mul_inv_cancel₀]
+              · simp only [mul_one]
+              · simp only [ne_eq, Nat.cast_eq_zero]
+                grind
+            · grind
+          rw [this]
+          unfold R2
+          simp only [add_assoc]
+
+lemma analyticOrderAt_eq_nat_iff_iteratedDeriv_eq_zero :
+  ∀ z₀ n (f : ℂ → ℂ) (_ : AnalyticAt ℂ f z₀) (_ : analyticOrderAt f z₀ ≠ ⊤),
+    (∀ k < n, deriv^[k] f z₀ = 0) ∧ (deriv^[n] f z₀ ≠ 0) ↔ analyticOrderAt f z₀ = n := by
+  intros z₀ n
+  induction n
+  · simp only [ne_eq, not_lt_zero', IsEmpty.forall_iff, implies_true,
+      Function.iterate_zero, id_eq, true_and, CharP.cast_eq_zero]
+    exact fun f hf ho ↦ Iff.symm (AnalyticAt.analyticOrderAt_eq_zero hf)
+  · rename_i n IH
+    intros f hf hfin
+    constructor
+    · intros H
+      obtain ⟨hz, hnz⟩:= H
+      have IH' := IH (deriv f) (AnalyticAt.deriv hf) ?_
+      · suffices analyticOrderAt (deriv f) z₀ = (n : ℕ) by
+          refine analyticOrderAt_eq_succ_iff_deriv_order_eq_pred f z₀ hf
+            (n + 1) (hz 0 (by omega)) this (by simp)
+        rw[← IH']
+        constructor
+        · intros k hk; exact hz (k + 1) (by omega)
+        · exact hnz
+      · specialize hz 0 (by omega)
+        obtain ⟨r, hr⟩ := (WithTop.ne_top_iff_exists).mp hfin
+        simp only [Function.iterate_zero, id_eq] at hz
+        have r0 : r > 0 := by
+          suffices analyticOrderAt f z₀ > 0 by
+            suffices @WithTop.some ℕ r > 0 by exact ENat.coe_lt_coe.mp this
+            rw [hr]
+            exact this
+          exact pos_of_ne_zero (analyticOrderAt_ne_zero.mpr ⟨hf, hz⟩)
+        have := Complex.analyticOrderAt_deriv_of_pos (n:= r) hf hr.symm (by omega)
+        rw [this]
+        exact ENat.coe_ne_top (r - 1)
+    · intros ho
+      constructor
+      · intros k hk
+        have : analyticOrderAt (deriv^[k] f) z₀ ≠ 0 := by
+          rw [(Complex.analyticOrderAt_iterated_deriv f hf k (n+1)
+            ho.symm (by omega) hk.le), @Nat.cast_ne_zero]
+          omega
+        rw [analyticOrderAt_ne_zero] at this
+        exact this.2
+      · have := Complex.analyticOrderAt_iterated_deriv f hf (n+1) (n+1)
+          ho.symm (by omega) (by omega)
+        simp only [Function.iterate_succ, Function.comp_apply, tsub_self,
+          CharP.cast_eq_zero] at this
+        rw [AnalyticAt.analyticOrderAt_eq_zero] at this
+        · assumption
+        · exact iterated_deriv hf (n + 1)
+
+lemma analyticOrderAt_eq_nat_imp_iteratedDeriv_eq_zero
+    z₀ (n : ℕ) (f : ℂ → ℂ) (hf : AnalyticAt ℂ f z₀) :
+  analyticOrderAt f z₀ = n → (∀ k < n, deriv^[k] f z₀ = 0) ∧ (deriv^[n] f z₀ ≠ 0) := fun h ↦
+  (analyticOrderAt_eq_nat_iff_iteratedDeriv_eq_zero z₀ n f hf (h.symm ▸ ENat.coe_ne_top n)).mpr h
+
+lemma le_analyticOrderAt_iff_iteratedDeriv_eq_zero : ∀ (n : ℕ) z₀
+  (f : ℂ → ℂ) (_ : AnalyticAt ℂ f z₀) (_ : analyticOrderAt f z₀ ≠ ⊤),
+  (∀ k < n, (deriv^[k] f) z₀ = 0) → n ≤ analyticOrderAt f z₀ := by
+    intros n z₀ f hf ho hkn
```

Comment:
```suggestion
lemma le_analyticOrderAt_iff_iteratedDeriv_eq_zero : ∀ (n : ℕ) z₀
  (f : ℂ → ℂ) (_ : AnalyticAt ℂ f z₀) (_ : analyticOrderAt f z₀ ≠ ⊤),
  (∀ k < n, (deriv^[k] f) z₀ = 0) → n ≤ analyticOrderAt f z₀ := by
    intros n z₀ f hf ho hkn
```
If you're `intro`ing, use normal variables instead of forall